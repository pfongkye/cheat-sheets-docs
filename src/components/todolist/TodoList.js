import React, { useState, useReducer } from "react";
import styled from "styled-components";
import Todo from "./Todo";

class TodoEntity {
  constructor(value) {
    this.value = value;
    this.Id = Date.now();
  }
}

const StyledTodoList = styled.div`
  color: blue;
`;

const initialState = [];
function reducer(state, action) {
  switch (action[0]) {
    case "ADD_TODO":
      return [action[1], ...state];
    case "COMPLETE_TODO":
      return [...state.filter((i) => i.Id !== action[1])];
    default:
      return state;
  }
}
export default function TodoList() {
  const [value, setValue] = useState("");
  const [todoItems, dispatch] = useReducer(reducer, initialState);

  function handleAdd() {
    dispatch(["ADD_TODO", new TodoEntity(value)]);
    setValue("");
  }

  function handleChange(e) {
    setValue(e.target.value);
  }

  function handleComplete(item) {
    dispatch(["COMPLETE_TODO", item.Id]);
  }

  return (
    <StyledTodoList>
      <h3>My Todo List</h3>
      <label>
        Todo: <input value={value} onChange={handleChange} />
      </label>
      <button onClick={handleAdd}>Add</button>
      <div>
        {todoItems.length > 0 && <span>My active todos:</span>}
        <div>
          {todoItems.map((item, i) => (
            <Todo key={`todo-${i}`} todo={item} onComplete={handleComplete} />
          ))}
        </div>
      </div>
    </StyledTodoList>
  );
}
