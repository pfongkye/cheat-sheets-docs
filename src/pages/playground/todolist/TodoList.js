import React from "react";
export default function TodoList() {
  return <div>My Todo List</div>;
}
