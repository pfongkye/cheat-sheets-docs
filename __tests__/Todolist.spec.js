import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";

import TodoList from "playground/todolist/TodoList";
describe("TodoList", () => {
  it("should display TodoList", () => {
    const { getByText } = render(<TodoList />);

    expect(getByText("My Todo List")).toBeInTheDocument();
  });

  it("should add a todo item and clear input value", () => {
    const { getByText, getByLabelText } = render(<TodoList />);
    const input = getAddTodoInput(getByLabelText);

    expectAddTodo(input, "My first todo", getByText).toBeInTheDocument();
    expect(input.value).toEqual("");
  });

  it("should add two todo items", () => {
    const { getByText, getByLabelText } = render(<TodoList />);
    const input = getAddTodoInput(getByLabelText);

    expectAddTodo(input, "My first todo", getByText).toBeInTheDocument();
    expectAddTodo(input, "My second todo", getByText).toBeInTheDocument();
  });

  it("should not show active todos when no todo", () => {
    const { queryByText } = render(<TodoList />);

    expect(queryByText("My active todos:")).not.toBeInTheDocument();
  });

  it("should show active todos when todo is added", () => {
    const { getByText, getByLabelText } = render(<TodoList />);

    const input = getAddTodoInput(getByLabelText);

    addTodo(input, "My first todo", getByText);
    expect(getByText("My active todos:")).toBeInTheDocument();
  });

  it("should complete one todo", async () => {
    const todo = "my todo";
    const { getByText, getByLabelText, queryByText } = render(<TodoList />);

    const todoInput = getAddTodoInput(getByLabelText);

    addTodo(todoInput, todo, getByText);

    completeTodo(getByLabelText, todo);

    await waitFor(() => {
      expect(queryByText("My active todos:")).not.toBeInTheDocument();
    });
  });

  it("should complete second todo and leave first one active", async () => {
    const firstTodo = "first todo",
      secondTodo = "second todo";

    const { getByText, getByLabelText, queryByText } = render(<TodoList />);

    const todoInput = getAddTodoInput(getByLabelText);

    addTodo(todoInput, firstTodo, getByText);
    addTodo(todoInput, secondTodo, getByText);

    completeTodo(getByLabelText, secondTodo);

    await waitFor(() => {
      expect(queryByText(secondTodo)).not.toBeInTheDocument();
      expect(getByText(firstTodo)).toBeInTheDocument();
    });
  });

  it("should complete both todos independently even if they have same value", async () => {
    const duplicatedTodo = "duplicated todo";

    const { getByText, getByLabelText, queryByText, getAllByLabelText } = render(<TodoList />);

    const todoInput = getAddTodoInput(getByLabelText);

    addTodo(todoInput, duplicatedTodo, getByText);
    addTodo(todoInput, duplicatedTodo, getByText);

    //since two todos of same value, we need to get all the todos...
    const todos = getAllByLabelText(duplicatedTodo, { selector: "input" });
    fireEvent.click(todos[0]);

    //afterwards we need to query again and not use the previous todos since DOM changed.
    completeTodo(getByLabelText, duplicatedTodo);

    await waitFor(() => {
      expect(queryByText(duplicatedTodo)).not.toBeInTheDocument();
    });
  });
});

function completeTodo(getByLabelText, todo) {
  fireEvent.click(getByLabelText(todo, { selector: "input" }));
}

function getAddTodoInput(getByLabelText) {
  return getByLabelText("Todo:", { selector: "input" });
}

function expectAddTodo(input, value, getByText) {
  addTodo(input, value, getByText);
  return expect(getByText(value));
}

function addTodo(input, value, getByText) {
  fireEvent.change(input, { target: { value } });
  fireEvent.click(getByText("Add"));
}
