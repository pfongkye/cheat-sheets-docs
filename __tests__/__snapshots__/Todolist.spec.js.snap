// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList should match snapshot 1`] = `
<div
  className="sc-AxirZ cOKHlg"
>
  <h3>
    My Todo List
  </h3>
  <label>
    Todo: 
    <input
      onChange={[Function]}
      value=""
    />
  </label>
  <button
    onClick={[Function]}
  >
    Add
  </button>
  <div>
    <div />
  </div>
</div>
`;
